core/templates/services/html-parsing.service.ts

```diff
@@ -0,0 +1,58 @@
+// Copyright 2023 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for handling HTML parsing
+ * operations focused on image tag detection and comparison.
+ */
+
+import {Injectable} from '@angular/core';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class HtmlParsingService {
+  private readonly IMAGE_TAG_REGEX =
+    /<oppia-noninteractive-image\b[^>]*?filepath-with-value=/g;
+
+  constructor() {}
+
+  /**
+   * Counts the number of <oppia-noninteractive-image> tags in the provided HTML string.
+   * @param {string} htmlString - The HTML string to analyze.
+   * @returns {number} The number of image tags found.
+   */
+  countImageTags(htmlString: string): number {
+    if (!htmlString || typeof htmlString !== 'string') {
+      return 0;
+    }
+
+    const imageTags = htmlString.match(this.IMAGE_TAG_REGEX);
+    return imageTags ? imageTags.length : 0;
+  }
+
+  /**
+   * Compares the original and updated HTML strings to determine if the number
+   * of images is mismatched.
+   * @param {string} originalHtml - The original HTML string with image tags.
+   * @param {string} updatedHtml - The updated (edited) HTML string with image tags.
+   * @returns {boolean} True if the number of image tags in the updated HTML
+   * does not match the original HTML.
+   */
+  isImageCountMismatched(originalHtml: string, updatedHtml: string): boolean {
```

Comment:
This has nothing to do with an "html parsing service". The logic here should stay outside this service.